core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -295,12 +314,63 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
         m_vector_length(verify_requested_vector_length(vector_length_request)),
         m_team_scratch_size{0, 0},
         m_thread_scratch_size{0, 0},
-        m_chunk_size(32) {
+        m_chunk_size(Impl::cuda_default_chunk_size),
+        m_tune_team(true),
+        m_tune_vector(false) {
     // Make sure league size is permissible
-    if (league_size_ >= int(Impl::cuda_internal_maximum_grid_count()))
+    if (league_size_ >= int(Impl::cuda_internal_maximum_grid_count())) {
+      Impl::throw_runtime_exception(
+          "Requested too large league_size for TeamPolicy on Cuda execution "
+          "space.");
+    }
+  }
+
+  /** \brief  Specify league size, request team size */
+  TeamPolicyInternal(const execution_space space_, int league_size_,
+                     const Kokkos::AUTO_t& /* team_size_request */,
+                     const Kokkos::AUTO_t& /* vector_length_request */
+                     )
+      : m_space(space_),
+        m_league_size(league_size_),
+        m_team_size(-1),
+        m_vector_length(-1),
+        m_team_scratch_size{0, 0},
+        m_thread_scratch_size{0, 0},
+        m_chunk_size(Impl::cuda_default_chunk_size),
+        m_tune_team(true),
+        m_tune_vector(true) {
+    // Make sure league size is permissible
+    if (league_size_ >= int(Impl::cuda_internal_maximum_grid_count())) {
       Impl::throw_runtime_exception(
           "Requested too large league_size for TeamPolicy on Cuda execution "
           "space.");
+    }
+  }
+
+  /** \brief  Specify league size, request team size */
+  TeamPolicyInternal(const execution_space space_, int league_size_,
+                     int team_size_request, const Kokkos::AUTO_t&)
+      : m_space(space_),
+        m_league_size(league_size_),
+        m_team_size(team_size_request),
+        m_vector_length(-1),
+        m_team_scratch_size{0, 0},
+        m_thread_scratch_size{0, 0},
+        m_chunk_size(Impl::cuda_default_chunk_size),
+        m_tune_team(false),
+        m_tune_vector(true) {
+    // Make sure league size is permissible
+    if (league_size_ >= int(Impl::cuda_internal_maximum_grid_count())) {
+      Impl::throw_runtime_exception(
+          "Requested too large league_size for TeamPolicy on Cuda execution "
+          "space.");
+    }
+
+    if (team_size_request >= Impl::cuda_max_hierarchical_parallelism) {
+      Impl::throw_runtime_exception(
+          std::string("Kokkos::TeamPolicy< Cuda > the team size is too large. "
+                      "Team size x vector length must be smaller than 1024."));
+    }
   }
 
   TeamPolicyInternal(int league_size_, int team_size_request,
```

Comment:
You could also treat -1 generally as as the "AUTO team size" which would allow you to have all consturctors delegate to the one which has ExecSpace, int team_size, int vector_length as arguments and reduce code duplication that way.

Reply:
I think I like both of these (the first: definitely, the second: probably). Let me try both and see if things break

Reply:
It appears this works